### ISSUE TITLE
DDC allows unsound abstract override

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#30568]
Consider the following code:

```dart
class A {
  void foo() {}
}
class B extends A {
  void foo([x]);
}
void f(B b) {
  b.foo(42);
}
main() {
  f(new B());
}
```

This is unsound, since it leads to `A.foo` being callable via the interface `B.foo`, meaning that it might be called with an unexpected argument.  But DDC compiles this code without issuing any errors, and silently ignores the argument at runtime.

A similar problem occurs here:

```dart
class A {
  void foo() {}
}
abstract class B extends A {
  void foo([x]);
}
class C extends B {}
void f(B b) {
  b.foo(42);
}
main() {
  f(new C());
}
```

In this case, class `B` is sound, since it's abstract, so it doesn't introduce a way for `A.foo` to be called via the interface `B.foo`.  But class `C` is unsound, since it is concrete and fails to override `foo`, so allows `A.foo` to be called via the interface `B.foo`.  Again DDC accepts this code without issuing any errors.

Note that soundness can be recovered if `C` overrides `foo`, e.g. the following is ok from a soundness perspective:

```dart
class A {
  void foo() {}
}
abstract class B extends A {
  void foo([x]);
}
class C extends B {
  void foo([x]) {}
}
void f(B b) {
  b.foo(42);
}
main() {
  f(new C());
}
```

(It's a weird corner case though; @lrhn has indicated that it's [undecided](https://codereview.chromium.org/2998383002/#msg8) whether this will be allowed in Dart 2.0)

It's also worth noting that if we move `B.foo` to an interface, the behavior of DDC/analyzer changes.  I.e. if we replace:

```dart
(abstract) class B extends A {
  void foo([x]);
}
```
with
```dart
abstract class I {
  void foo([x]);
}
(abstract) class B extends A implements I {}
```

then DDC/analyzer issues compile errors for all three examples.  Whatever we decide to do, we should make sure that DDC/analyzer behaves consistently regardless of whether `B` has an abstract override of `foo` with an optional argument, or brings in that interface using an `implements` clause.

I'll be submitting a CL with test cases demonstrating this bug shortly.

**Comment by bwilkerson** [comment]
@stereotype441 Is this still an issue, or should we close this?

**Comment by stereotype441** [comment]
The test cases still fail, so I think we need to leave the issue open:

```
paulberry@paulberry-linux-2:~/dart1/sdk$ git grep -n 30568
tests/language_2/language_2_analyzer.status:7:abstract_override_adds_optional_args_concrete_subclass_test: MissingCompileTimeError # Issue #30568
tests/language_2/language_2_analyzer.status:8:abstract_override_adds_optional_args_concrete_test: MissingCompileTimeError # Issue #30568
tests/language_2/language_2_analyzer.status:9:abstract_override_adds_optional_args_supercall_test: MissingCompileTimeError # Issue #30568
tests/language_2/language_2_analyzer.status:1230:additional_interface_adds_optional_args_test: CompileTimeError # Issue #30568
tests/language_2/language_2_dartdevc.status:13:additional_interface_adds_optional_args_test: CompileTimeError # Issue #30568
tests/language_2/language_2_dartdevc.status:1034:abstract_override_adds_optional_args_concrete_subclass_test: MissingCompileTimeError # Issue #30568
tests/language_2/language_2_dartdevc.status:1035:abstract_override_adds_optional_args_concrete_test: MissingCompileTimeError # Issue #30568
tests/language_2/language_2_dartdevc.status:1036:abstract_override_adds_optional_args_supercall_test: MissingCompileTimeError # Issue #30568
```

**Comment by lrhn** [comment]
It is now decided that an abstract class is allowed to not implement its interface in Dart 2.
It must have a valid and unambiguous interface, but it doesn't have to have an implementation that is compatible with that interface. Super-calls are checked against that actual implementation signature they are trying to call, not the interface of the super-class (which may or may not match).

**Comment by stereotype441** [comment]
@lrhn thanks for the follow-up.  It looks like the expectations in tests/language_2/abstract_override_adds_optional_args_*.dart are consistent with this decision.

**Comment by MichaelRFairhurst** [comment]
This seems like a fairly big deal to me, I'm adding this to dart2.1

**Comment by MichaelRFairhurst** [comment]
This looks very close to being fixed in Fasta: https://github.com/dart-lang/sdk/issues/32014

**Comment by stereotype441** [comment]
Fixed in c44fa48f8fc848c55541a5b7febf97d0c3846989
